src/hotspot/share/runtime/sharedRuntime.cpp

```diff
@@ -996,11 +996,12 @@ JRT_ENTRY_NO_ASYNC(void, SharedRuntime::register_finalizer(JavaThread* current,
 JRT_END
 
 jlong SharedRuntime::get_java_tid(Thread* thread) {
-  if (thread != NULL) {
-    if (thread->is_Java_thread()) {
-      oop obj = JavaThread::cast(thread)->threadObj();
-      return (obj == NULL) ? 0 : java_lang_Thread::thread_id(obj);
-    }
+  if (thread != NULL && thread->is_Java_thread()) {
+    Thread* current = Thread::current();
+    guarantee(current != thread || !JavaThread::cast(thread)->is_gc_barrier_detached(),
+              "current cannot touch oops after its GC barrier is detached.");
+    oop obj = JavaThread::cast(thread)->threadObj();
```

Comment:
I think the oop-touching-safety check should be done in `threadObj()` itself so that all callers are protected.

Reply:
The placement in SharedRuntime::get_java_tid() is because that is the failure
mode that this bug is about. My next stress testing experiment will be to put
a similar guarantee() in threadObj(), but I expect there to be more fan-out
from that placement.

This fix is tightly focued on the bug as it is reported because I plan to fix
it in JDK19.

Reply:
Yeah ... hmmm ... threadObj is still the better place for any safety check ... arguably it should be even deeper in OopHandle.

I'm not sure what fanout you are concerned about. If we find there are other invalid oop accesses then we would want to fix them in 19 - no?

Reply:
I want to fix this bug in JDK19 since it is a blocker for @zhengyu123. I want to
keep this fix small and focused on the exact crash associated with the bug.

Yes, threadObj() or deeper is a better place. No argument, but I plan to do that
as a new hunt and I expect there to be many problems and I don't want to hold
up this fix for additional hunting. If I find more issues, then we'll determine how
to get them fixed in JDK19.

Reply:
New bug:
[JDK-8289091](https://bugs.openjdk.org/browse/JDK-8289091) move oop safety check from SharedRuntime::get_java_tid() to JavaThread::threadObj()